### PR TITLE
Implement voiding covers

### PR DESCRIPTION
**Pull request by PrototypeTrousers**
This PR implements 2 Fluid voiding covers and 2 Item voiding covers
Theres no rate limiting on them.
Both basic accepts filtering.
Advanced ones can have amount set on the filter, like the robotic arms and fluid regulator
Still missing recipes for them.

### CODE REVIEW

File: src/main/java/gregtech/common/covers/CoverFluidVoidingAdvanced.java [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+    }
+
+    /**
+     * Performs one tick worth of Keep Exact behavior.
+     *
+     * @param sourceHandler source(s) to move fluids from
+     * @param fluidFilter   a predicate which determines what fluids may be moved
+     * @param keepAmount    the desired amount in milliBuckets of a particular fluid in the destination
+     */
+    protected void voidOverflow(final IFluidHandler sourceHandler,
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/common/covers/CoverFluidVoidingAdvanced.java:58]: One thing that would be nice there would the the ability to set the `keepAmount` separately for different fluids. This can already be done for items, by changing the stack size of the item in the filter, but this cannot be done for fluids at all, so you end up applying just 1 number to all fluids, even though you may have wanted to limit the fluids to different amounts.

This was not really possible in CE, as it did not have the fluidstack stack size display in the slot, so you would never know what you were limiting the fluids to, but since we have that feature in CEu, the ability to set different fluid amounts in the fluid filter would be very appreciated. (If done for the fluid voiding cover, it should also be done for the Fluid Regulator, maybe at another time though)

File: src/main/java/gregtech/common/covers/CoverFluidRegulator.java [hunk 2]
```diff
@@ -37,6 +39,12 @@
     public CoverFluidRegulator(ICoverable coverHolder, EnumFacing attachedSide, int tier, int mbPerTick) {
         super(coverHolder, attachedSide, tier, mbPerTick);
         this.transferMode = TransferMode.TRANSFER_ANY;
+        this.fluidFilter = new FluidFilterContainer(this, this::shouldShowTip, maxFluidTransferRate * 100);
+    }
+
+    @Override
+    protected boolean shouldShowTip() {
+        return transferMode != TransferMode.TRANSFER_ANY;
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/common/covers/CoverFluidRegulator.java:47]: This Tio needs to be updated to state that holding control and shift changes the increment amount by 1000mb. I could not find where this top was defined, so I am putting it here

File: src/main/java/gregtech/common/covers/CoverFluidRegulator.java [hunk 3]
```diff
@@ -86,48 +97,53 @@ protected int doTransferExact(int transferLimit, IFluidHandler sourceHandler, IF
 
     /**
      * Performs one tick worth of Keep Exact behavior.
+     *
      * @param transferLimit the maximum amount in milliBuckets that may be transferred in one tick
      * @param sourceHandler source(s) to move fluids from
-     * @param destHandler destination(s) to move fluids to
-     * @param fluidFilter a predicate which determines what fluids may be moved
-     * @param keepAmount the desired amount in milliBuckets of a particular fluid in the destination
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/common/covers/CoverFluidRegulator.java:93]: Why was this line removed if the variable was still passed as a parameter?

File: src/main/java/gregtech/common/covers/CoverFluidRegulator.java [hunk 4]
```diff
@@ -86,48 +97,53 @@ protected int doTransferExact(int transferLimit, IFluidHandler sourceHandler, IF
...
 
         int transferred = 0;
-        for(FluidStack fluidStack : sourceFluids.keySet()) {
-            if(transferred >= transferLimit)
+        for (FluidStack fluidStack : sourceFluids.keySet()) {
+            if (transferred >= transferLimit)
                 break;
 
+            if (this.fluidFilter.getFilterWrapper().getFluidFilter() != null) {
+                keepAmount = this.fluidFilter.getFilterWrapper().getFluidFilter().getFluidTransferLimit(fluidStack);
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/common/covers/CoverFluidRegulator.java:127]: Should this be the maximum/minimum between the transfer limit and the passed keep amount?

**Review by ALongStringOfNumbers — CHANGES_REQUESTED** [review]
The scrolling for fluid amount in the advanced fluid cover seems to be broken, I dragged in a bucket of water from JEI, which occupied the slot with 1000mb. I scrolled once, which changed to 1 mb, I scrolled a second time, which changed to -100mb. Then it would change between 1 and -100 every time I scrolled after that.

Related minorly to the above, the fluid amount would change to 1mb when I left clicked the fluid in the filter, and -100mb when I right clicked the fluid in the filter.

In other fluid filter applications, Right clicking clears the slot, which needs to happen for this filter, because otherwise there is no way to clear the filter in the cover.

Could the tooltip on the voiding covers say "Needs to be activated with Soft Hammer after placement"?

File: src/main/java/gregtech/common/covers/CoverFluidRegulator.java (hunk 4 above)
Comment (PrototypeTrousers) [reply src/main/java/gregtech/common/covers/CoverFluidRegulator.java:127]: keepAmount should be the global if the mode is not keep exact/supply or overflown

File: src/main/java/gregtech/common/covers/CoverFluidRegulator.java (hunk 2 above)
Comment (PrototypeTrousers) [reply src/main/java/gregtech/common/covers/CoverFluidRegulator.java:46]: the tooltip is shown on `drawInForeground` on `PhantomFluidWidget`

File: src/main/java/gregtech/common/covers/CoverItemVoidingAdvanced.java [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+            if (sourceStack.isEmpty()) {
+                continue;
+            }
+            if (!itemFilterContainer.testItemStack(sourceStack)) {
+                continue;
+            }
+            myItemHandler.extractItem(srcIndex, Integer.MAX_VALUE, false);
+        }
+    }
+
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/common/covers/CoverItemVoidingAdvanced.java:61]: Is this needed? It looks to be the exact same as `CoverItemVoiding#voidAny`, which this filter has access to

File: src/main/java/gregtech/common/covers/CoverFluidRegulator.java [hunk 6]
```diff
@@ -315,12 +334,20 @@ public void writeToNBT(NBTTagCompound tagCompound) {
...
         tagCompound.setInteger("TransferAmount", transferAmount);
+        tagCompound.setTag("filterv2", new NBTTagCompound());
     }
 
     @Override
     public void readFromNBT(NBTTagCompound tagCompound) {
         super.readFromNBT(tagCompound);
         this.transferMode = TransferMode.values()[tagCompound.getInteger("TransferMode")];
+        //legacy NBT tag
+        if (!tagCompound.hasKey("filterv2") && tagCompound.hasKey("TransferAmount")) {
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/common/covers/CoverFluidRegulator.java:345]: I Notice that this is the only cover with this legacy check, but do the new covers even serialize with this Tag? I don't see them overriding the Pump/Conveyor write/read toNBT methods, nor do I see any changes in the Fluid Filter serialization tag names

**Review by ALongStringOfNumbers — COMMENTED** [review]
I saw some weird behavior with the advanced fluid voiding cover.

I placed down the cover on a fluid heater, which was constantly producing steam. I set the mode to Void Overflow, with a rate of 14999L/s, but did not place a fluid filter into the cover.

When I activated the cover, it would void steam disregarding whitelist/blacklist mode. 

What would be your expected behavior in this case? I would expect some difference between the whitelist and blacklist modes, or else maybe for the button to not appear unless a filter was inserted into the cover


For the Item Voiding cover, it does not pull from the non-virtualized inventory of a super chest, although that might be more on the super chest's side.

Some weird differences between the fluid cover and the item cover Void Overflow amount entries. The advanced item cover can only go up to 1024 in its text box, while the fluid cover can go up to `9999999999` in L/s mode. 

Also, when the fluid cover was in the above setup, switching to kL/s mode with the text in the box selected crashed with this crash: https://pastebin.com/Y7M92jEu
Although this may just be a TextField2 error

Could you also adjust the tooltip for `cover.fluid_regulator.transfer_mode.description` to mention the control+shift for a factor of 1000?

File: src/main/java/gregtech/common/covers/CoverItemVoidingAdvanced.java [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+    @Override
+    public ModularUI createUI(EntityPlayer player) {
+        WidgetGroup primaryGroup = new WidgetGroup();
+        primaryGroup.addWidget(new LabelWidget(10, 5, getUITitle()));
+
+        this.initFilterUI(20, primaryGroup::addWidget);
+
+        WidgetGroup filterGroup = new WidgetGroup();
+        filterGroup.addWidget(new CycleButtonWidget(91, 14, 75, 20,
```
[the remark this answers is not in the document]
Comment (PrototypeTrousers) [reply src/main/java/gregtech/common/covers/CoverItemVoidingAdvanced.java:103]: the grid on the item filter shows to the right of the filter slot and the grid on the fluid filter below.

File: src/main/java/gregtech/common/covers/CoverFluidRegulator.java (hunk 6 above)
Comment (PrototypeTrousers) [reply src/main/java/gregtech/common/covers/CoverFluidRegulator.java:345]: yeah this is actually only needed on the fluid regulator, since we dont have to migrate the voiding covers yet...
since this PR changes the behavior of already placed fluid filters, this tag is only used to mark if the amount previously set on the regulator as an amount to supply/ extract exact will also be set to the configured fluids on the filter

**Comment by PrototypeTrousers** [comment]
Okay.. the Blacklist mode on Overflow voiding is not making any sense, it acts like a blacklist void any. But im really not comfortable with switch the filter between white and blacklists without being explicit about that. 
Any ideas?

**Comment by ALongStringOfNumbers** [comment]
Honestly, I would suggest disabling the blacklist for Voiding Overflow mode, as it does not really make sense for the mode. The player has to specify the amount to void, which does not make sense then if there is a blacklist being applied to that amount.

**Review by ALongStringOfNumbers — APPROVED** [review]
:see_no_evil:
